Dirichlet family

micbalz opened the pull request:
Dear gamboostLSS-Team,

I have implemented a novel Dirichlet family which allows for fitting Dirichlet regression models via gradient boosting within your gamboostLSS package for a paper that is currently under review. One referree pointed out that it would be favorable to add the family to gamboostLSS. To this end, I adjusted the necessary file in my fork and conducted relevant tests.

I would be really happy if you would consider adding the Dirichlet family.

Kind regards, 
Michael

micbalz committed:
implement Dirichlet family

micbalz committed:
delete

micbalz committed:
fix typos

hofnerb commented:
@mayrandy, it seems that our devel branch is completely outdated and that Michael (@Tanzmarie) has used that branch to develop his family. I am not sure if one can still change the merge target to master and check for differences there? I am currently having issues reviewing the code as it is based on many changes from the devel branch. 

As a sidenote: We should consider what to do with the devel branch. It seems that we should either update it to the most current version or remove it completely to avoid issues. Maybe we can have a short call in the comming days?

hofnerb commented:
@Tanzmarie: The code is currently not ok. Please run R checks and remove all notes, warnings and errors related to your code. A quick check gives me the following issues:

- [ ] DirichletAlpha: no visible binding for global variable 'loss'
  Undefined global functions or variables:
   loss

- [ ] checking Rd line widths ... NOTE
Rd file 'families.Rd':
  \usage lines wider than 90 characters:
     DirichletLSS(K = NULL, a1=NULL, a2=NULL, a3=NULL, a4=NULL, a5=NULL, a6=NULL, a7=NULL, a8=NULL, a9=NULL,
                m1=NULL, m2=NULL, m3=NULL, m4=NULL, m5=NULL, m6=NULL, m7=NULL, m8=NULL, m9=NULL,
                s1=NULL, s2=NULL, s3=NULL, s4=NULL, s5=NULL, s6=NULL, s7=NULL, s8=NULL, s9=NULL,

     These lines will be truncated in the PDF manual.

- [ ] checking for code/documentation mismatches ... WARNING
Codoc mismatches from Rd file 'families.Rd':
DirichletLSS
  Code: function(K, a1 = NULL, a2 = NULL, a3 = NULL, a4 = NULL, a5 =
                 NULL, a6 = NULL, a7 = NULL, a8 = NULL, a9 = NULL, m1 =
                 NULL, m2 = NULL, m3 = NULL, m4 = NULL, m5 = NULL, m6 =
                 NULL, m7 = NULL, m8 = NULL, m9 = NULL, s1 = NULL, s2 =
                 NULL, s3 = NULL, s4 = NULL, s5 = NULL, s6 = NULL, s7 =
                 NULL, s8 = NULL, s9 = NULL, ...)
  Docs: function(K = NULL, a1 = NULL, a2 = NULL, a3 = NULL, a4 = NULL,
                 a5 = NULL, a6 = NULL, a7 = NULL, a8 = NULL, a9 = NULL,
                 m1 = NULL, m2 = NULL, m3 = NULL, m4 = NULL, m5 = NULL,
                 m6 = NULL, m7 = NULL, m8 = NULL, m9 = NULL, s1 = NULL,
                 s2 = NULL, s3 = NULL, s4 = NULL, s5 = NULL, s6 = NULL,
                 s7 = NULL, s8 = NULL, s9 = NULL, ...)
  Mismatches in argument default values:
    Name: 'K' Code:  Docs: NULL

- [ ] checking Rd \usage sections ... WARNING
Undocumented arguments in Rd file 'families.Rd'
  'K' 'a1' 'a2' 'a3' 'a4' 'a5' 'a6' 'a7' 'a8' 'a9' 'm1' 'm2' 'm3' 'm4'
  'm5' 'm6' 'm7' 'm8' 'm9' 's1' 's2' 's3' 's4' 's5' 's6' 's7' 's8' 's9'

    Functions with \usage entries need to have the appropriate \alias
    entries, and all their arguments documented.
    The \usage entries must correspond to syntactically valid R code.
    See chapter 'Writing R documentation files' in the 'Writing R
    Extensions' manual.

- [ ] checking for unstated dependencies in 'tests' ... WARNING
'library' or 'require' call not declared from: 'DirichletReg'

- [ ] Running the tests in 'tests/regtest-families.R' failed.
Last 13 lines of output:
  > n <- 150
  > p <- 10
  > 
  > x <- matrix(runif(p * n, 0,1), n)
  > 
  > x <- data.frame(x)
  > 
  > a1 <- exp(2.5*x[,1] - x[,2] + 3*x[,3]) 
  > a2 <- exp(2*x[,4] + 2*x[,5] - x[,6])
  > a3 <- exp(1.5*x[,7] -  1.5*x[,8] + x[,9])
  > A <- cbind(a1,a2,a3)
  > 
  > y <- rdirichlet(nrow(A),A)
  Error in rdirichlet(nrow(A), A) : could not find function "rdirichlet"
  Execution halted


If possible you could also use the master branch and add your code to that to make the pull request easier as discussed above. I am also wondering if you really need the lengthy lists of arguments (a1, ...) which I currently don't even understand given a lack of documentation. Maybe we can see if it possible to use a vector or list of arguments instead?

micbalz committed:
Merge branch 'boost-R:master' into master

micbalz committed:
run R checks, update documentation, remove notes

Tanzmarie commented:
Dear @hofnerb, thank you for the constructive suggestions and feedback. 

First, I would like to point out that I forked gamboostLSS from the up to the date master branch. Therefore, the Dirichlet family is implemented in the most recent version. I simply followed previous pull requests and thus chose the devel branch. Should I make a pull request for the master branch instead?

Second, I acknowledge the previously problems with the code and notes on the R checks. I updated and (hopefully) improved the code. I ran extensive R checks and removed all notes related to the implemented Dirichlet family. If the implementation has still problems that I missed then I will gladly address them.

micbalz commented:
Dear @hofnerb, I was actually able to change the pull request from the devel to the recent master branch. I hope this resolves any problem associated with the devel branch.

hofnerb commented:
I just had a look at the code. Overall the code looks much better! Thank you very much.
Some points still caught my attention:


  1.  The ngradient function accepts weights but does not seem to use them. Please check and fix (and write a test routine).
  2.  Other gamboostLSS families use the stabilize argument. This argument normalizes/stabilizes the partial negative gradients to make them comparable (see also the manual for Families). Does this make sense for your Dirichlet family? If not, one could consider to “capture” this in the code and/or mention in the manual. We need to check for consistency with other Famlies(), though.
  3.  I need to check the Families() object returned by the DirichletLSS function in detail. Currently, it seems to lack some arguments expected in Families(), i.e., qfun (quantile function. This function can for example be used to compute (marginal) prediction intervals. See predint), name and potentially stabilization as discussed above. Consider to add.
  4.  Documentation of the functions is still not very extensive. What does the code actually do, especially the “eval(parse(…))” parts? Please document these more complex steps sufficiently to make it possible to maintain the code in the future.
  5.  There is no longer a possibility to define offsets manually. Is this desirable?
  6.  Consider to add your paper to the references of the manual for DirichletLSS and cite it in the Details section if you want.

Von: Michael Balzer ***@***.***>
Gesendet: Montag, 20. Januar 2025 12:00
An: boost-R/gamboostLSS ***@***.***>
Cc: Hofner, Benjamin ***@***.***>; Mention ***@***.***>
Betreff: Re: [boost-R/gamboostLSS] Dirichlet family (PR #60)


Dear @hofnerb<https://github.com/hofnerb>, I was actually able to change the pull request from the devel to the recent master branch. I hope this resolves any problem associated with the devel branch.

—
Reply to this email directly, view it on GitHub<https://github.com/boost-R/gamboostLSS/pull/60#issuecomment-2602106488>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACDKCMBT36UQEHHXM5JDO5D2LTJMXAVCNFSM6AAAAABUGDJRAGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMBSGEYDMNBYHA>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>
[ { ***@***.***": "http://schema.org", ***@***.***": "EmailMessage", "potentialAction": { ***@***.***": "ViewAction", "target": "https://github.com/boost-R/gamboostLSS/pull/60#issuecomment-2602106488", "url": "https://github.com/boost-R/gamboostLSS/pull/60#issuecomment-2602106488", "name": "View Pull Request" }, "description": "View this Pull Request on GitHub", "publisher": { ***@***.***": "Organization", "name": "GitHub", "url": "https://github.com" } } ]

micbalz committed:
adjust families, add comments, add tests

micbalz committed:
fix notes in checks

micbalz commented:
Thank you very much for your detailed feedback, @hofnerb. I really appreciate your insights and have addressed your points as follows.

First, thanks for catching that mistake regarding the weights in the ngradient function. I’ve now corrected this by ensuring that the declared weights are properly used in the functions, and I have tested them to confirm everything works as expected.

Regarding the stabilization argument, I’ve added it and also implemented some test routines to verify its effect. For the DirichletLSS function, I’ve made adjustments so that both the function name and stabilization argument are now included. However, I’d like to point out that the Dirichlet distribution is a multivariate distribution over a simplex, meaning its quantiles are not uniquely defined in the same way as for univariate distributions. Because of this, I aligned the approach with the loglogLSS family, which also apparently does not utilize qfun.

To improve clarity, I have rewritten the eval() parts and provided extensive comments, making the documentation more transparent and easier to maintain in the future. Also, choosing offset values manually is not desirable in this case, so I have accounted for that accordingly.

Finally, I appreciate your suggestion regarding the paper. Since it is still under revision and has not been published yet, I will hold off on adding it to the details section for now, but I will include it at the appropriate time.

micbalz committed:
fix errors

hofnerb committed:
Remove testthat from suggests as it is not needed.

hofnerb commented:
It seems that all is fine now. I didn't check that the code does what it should do but all other issues seem resolved. I will now merge the pull request to gamboostLSS master branch and will try to make a CRAN release as soon as possible.

hofnerb commented:
@micbalz The new package is now on CRAN.
